hermes/probe/create/create_file.go

```diff
@@ -0,0 +1,160 @@
+		return &RandomFile{}, fmt.Errorf("invalid argument: RandomFile.Size = %d; want 0 < size <= %d", fileSize, maxFileSizeBytes)
+	}
+	return &RandomFile{ID: int64(fileID), Size: fileSize}, nil
+}
+
+func (f *RandomFile) CheckSum() ([]byte, error) {
+	r := f.NewReader()
+	h := sha1.New()
+	if _, err := io.Copy(h, r); err != nil {
+		return nil, fmt.Errorf("io.Copy: %v", err)
+	}
+	return h.Sum(nil), nil
+}
+
+func (f *RandomFile) FileName() (string, error) {
+	checksum, err := f.CheckSum()
+	if err != nil {
+		return "", err
+	}
+	return fmt.Sprintf("Hermes_%02d_%x", f.ID, checksum), nil
+}
+
+func CreateFile(ctx context.Context, target *probe.Target, fileID int32, fileSize int, client stiface.Client, logger *logger.Logger) error {
+	f, err := NewRandomFile(int64(fileID), fileSize)
+	if err != nil {
+		return err
+	}
+	fileName, err := f.FileName()
+	if err != nil {
+		return err
+	}
+	target.Journal.Intent = &journalpb.Intent{FileOperation: journalpb.Intent_CREATE, Filename: fileName}
+	if _, ok := target.Journal.Filenames[fileID]; ok {
+		var status m.ExitStatus
+		status = m.UnknownFileFound
+		return fmt.Errorf("CreateFile(ID: %d).%v could not create file as file with this ID already exists", fileID, status)
+	}
+	r := f.NewReader()
+	start := time.Now()
+	bucketName := target.Target.GetBucketName()
+	wc := client.Bucket(bucketName).Object(fileName).NewWriter(ctx)
+	if _, err = io.Copy(wc, r); err != nil {
+		var status m.ExitStatus
+		switch err {
+		case storage.ErrBucketNotExist:
+			status = m.BucketMissing
+		default:
+			status = m.ProbeFailed
+		}
+		target.LatencyMetrics.APICallLatency[m.APICreateFile][status].Metric("hermes_api_latency_s").AddFloat64(time.Now().Sub(start).Seconds())
+		return fmt.Errorf("CreateFile(id: %d).%v: could not create file %s: %w", fileID, status, fileName, err)
+	}
+	if err := wc.Close(); err != nil {
+		target.LatencyMetrics.APICallLatency[m.APICreateFile][m.WriterCloseFailed].Metric("hermes_api_latency_s").AddFloat64(time.Now().Sub(start).Seconds())
+		return fmt.Errorf("Writer.Close: %v", err)
+	}
+	target.LatencyMetrics.APICallLatency[m.APICreateFile][m.Success].Metric("hermes_api_latency_s").AddFloat64(time.Now().Sub(start).Seconds())
+	prefix := fileName[0:9]
+	query := &storage.Query{Prefix: prefix}
+	start = time.Now()
+	objIter := client.Bucket(bucketName).Objects(ctx, query)
+	end := time.Now()
+	obj, err := objIter.Next()
+	if err != nil {
+		target.LatencyMetrics.APICallLatency[m.APIListFiles][m.FileMissing].Metric("hermes_api_latency_s").AddFloat64(end.Sub(start).Seconds())
+		return fmt.Errorf("CreateFile check failed: %w", err)
+	}
+	if obj.Name != fileName {
+		fmt.Errorf("CreateFile check failed expected file name present %v got %v", fileName, obj.Name)
+	}
+	target.LatencyMetrics.APICallLatency[m.APIListFiles][m.Success].Metric("hermes_api_latency_s").AddFloat64(end.Sub(start).Seconds())
+
+	target.Journal.Filenames[fileID] = fileName
+	if logger != nil {
```

Comment:
I would put a contract in the docstring the caller must pass in a valid logger. Then there's no need for these checks.

Reply:
In the test file the logger passed is nil because w don't want to bring up cloudprober hence the check in the code. Adding doc strings is left as a TODO

Reply:
You could pass in a fake logger that just calls fmt.Println().

Reply:
replaced! added the fakeLogger interface to the fakegcs file and package for now but do you think it should be its own separate package perhaps fakelogger?

Reply:
I think it's cleaner to put it in a separate package.